docs: Add an introduction to actors

llucax opened the pull request:
This new document introduces the actor programming model and how to use it in Frequenz SDK.

Fixes #648.

llucax committed:
Add a macro to easily display a code annotation

This is to encourage users to expand annotations, which might not be
that obvious otherwise.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Make code annotations numbered

This allows to have a read order for annotations.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add an introduction to actors

This document introduces the actor programming model and how to use it in
Frequenz SDK.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Based on #667, ignore the first 4 commits.

daniel-zullo-frequenz commented on docs/intro/actors.md:
````diff
@@ -0,0 +1,148 @@
+    asyncio.run(main())
+```
+
+1. We define 2 actors: `Actor1` and `Actor2` that will just forward message
````
```suggestion
1. We define 2 actors: `Actor1` and `Actor2` that will just forward a message
```

daniel-zullo-frequenz commented on docs/intro/actors.md:
```diff
@@ -0,0 +1,148 @@
+   [senders][frequenz.channels.Sender] and
+   [receivers][frequenz.channels.Receiver] from the channels.
+
+6. We schedule the [sending][frequenz.channels.Sender.send] of message `Hello`
+   to `Actor1` via `input_channel`.
```
```suggestion
6. We schedule the [sending][frequenz.channels.Sender.send] of the message
   `Hello` to `Actor1` via `input_channel`.
```

daniel-zullo-frequenz reviewed (commented):
I have a couple of cosmetic comments to check for. LGTM!

llucax commented:
Did you review all the commits or only what wasn't in #667? If you reviewed everything maybe you can approve #667 too :D

llucax committed:
Move `_slugify()` to the top-level

This function doesn't need to access the `env` object, so it can be
defined at the top-level.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Investigating an issue with numbered annotations rendering wrongly...

llucax commented:
Seems to be an upstream bug: https://github.com/squidfunk/mkdocs-material/issues/6042, so I guess we can proceed hoping it will be fixed.

shsms reviewed (approved):
This looks good to me, but at some point, we should write additional sections on requirements, best practices, composition patterns, etc.

llucax commented:
If you have any specific suggestions I can try to add some of that, it is weird, when I'm coding or reading issues I see a lot of things that it would be good to have in docs, but when I sit down to write docs I get a writers block and I don't know what to put in there :laughing:

llucax commented:
But I think I will merge this since it is approved so we can keep moving, I can write up any suggestions you make later in a separate PR.
